File: src/238. Product of Array Except Self.cpp
Problem link: https://leetcode.com/problems/product-of-array-except-self/

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> left_product(n), right_product(n), final_product(n);
        right_product[0] = 1;
        left_product[n - 1] = 1;
        for (int i = 0; i < n - 1; i++) {
            right_product[i + 1] = right_product[i] * nums[i];
            left_product[n - i - 2] = left_product[n - i - 1] * nums[n - i - 1];
        }
        for (int i = 0; i < n; i++) {
            final_product[i] = left_product[i] * right_product[i];
        }
        return final_product;
    }
};

File: Leetcode/238. Product of Array Except Self.java
/* Problem Link: https://leetcode.com/problems/product-of-array-except-self/ */

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] product_from_left = new int[nums.length];
        int[] product_from_right = new int[nums.length];
        int[] ans = new int[nums.length];

        for (int i = 0; i < nums.length; i++) {
            if (i == 0) {
                product_from_left[i] = nums[i];
            } else {
                product_from_left[i] = product_from_left[i - 1] * nums[i];
            }
        }

        for (int i = nums.length - 1; i >= 0; i--) {
            if (i == nums.length - 1)
                product_from_right[i] = nums[i];
            else
                product_from_right[i] = product_from_right[i + 1] * nums[i];
            try {
                ans[i] = product_from_right[i + 1] * product_from_left[i - 1];
            } catch (Exception e) {
                try {
                    ans[i] = product_from_right[i + 1];
                } catch (Exception f) {
                    ans[i] = product_from_left[i - 1];
                }
            }
        }
        return ans;
    }
}
